docs/elastic.rst

```diff
@@ -0,0 +1,358 @@
+.. inclusion-marker-start-do-not-remove
+
+Elastic Horovod
+===============
+
+
+Elastic training enables Horovod to scale up and down the number of workers dynamically at runtime, without
+requiring a restart or resuming from checkpoints saved to durable storage. With elastic training, workers can come
+and go from the Horovod job without interrupting the training process.
+
+
+When to use elastic training
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- You are running an `autoscaling <https://en.wikipedia.org/wiki/Autoscaling>`__ job that may acquire more resources for training over time.
+- Your job is running on preemptable or spot instances that may come and go with little warning.
+- Your nodes are unreliable and you want your job to continue training if some of the hosts fail.
+
+
+Requirements
+~~~~~~~~~~~~
+
+- Python >= 3.6
+- TensorFlow >= 1.15 or PyTorch >= 1.0
+- Horovod >= 0.20.0 with Gloo support (install Horovod using ``HOROVOD_WITH_GLOO=1`` to ensure it is installed)
+- A way to discover available hosts at runtime
+
+
+Modifying the training script with State Synchronization
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The biggest difference when moving from normal distributed training to elastic training is the need to track and synchronize
+state among the workers as workers are added or removed from the job.
+
+To enable elastic training, make the following changes to your training script:
+
+1. Wrap your main training process (everything following initialization) in a function decorated with ``hvd.elastic.run``.
+
+   The first argument to this decorated function should be an instance of ``hvd.elastic.State``.  Before executing the
+   decorated function, this state object will be synchronized across workers.  This ensures that workers that were
+   newly added, as well as workers that might have inconsistent state, all share the same state before training begins.
+
+   Because the sync function uses collective ops, and upon worker add the active workers will not reset from before this
+   function, *no Horovod collective ops (broadcast, allreduce, allgather, etc.) can be called before this function*.
+
+2. Place all variables that need to be kept in sync between worker replicas (model parameters, optimizer state, epoch and batch numbers, etc.) into a ``hvd.elastic.State`` object.
+
+   Standard state implementations are provided for TensorFlow, Keras, and PyTorch.  However, it may be necessary in some cases to override
+   the base ``hvd.elastic.State`` object to handle broadcasting custom types.
+
+3. Periodically call ``state.commit()`` to backup a copy of your state in memory.
```

Comment:
Is it a requirement to run a distributed operation like `hvd.allgather` just before a commit, since commit is not failing when nodes dropped out of the cluster? Otherwise workers will successfully continue to perform `state.commit` after some nodes failed, and new nodes will start from states later than where the workers terminated, but state will not contain global (`allgather`) knowledge.

Reply:
I found that state.commit calls into

        self.save()
        self.check_host_updates()

where the latter performs a broadcast operation. Does that one fail if any node other then `rank == 0` dropped out of the cluster? Does this synchronize all workers?

Still, node `rank == 0` may have committed some batches that have not been done on other workers, and on reset those batches will not be re-done on other workers, they will be skipped. Is that a general problem or is this training-script specific and therefore needs to be considered on a per-script basis?

Reply:
If a failure occurs during broadcast / commit, then a `HorovodInternalError` should be raised, which will result in a reset event (this can happen if one of the other workers fail, as you suggested).  

Because Horovod synchronizes on the allreduce, we can feel confident that if we call commit at the end of the batch, we should not need to worry about workers being out of sync.  So it should not be possible for some workers to get ahead of others.  If a worker fails in between allreduce and commit, then we should still be okay, because the gradient updates of the failing worker should still be applied to the other non-failing workers.